cpp/dolfinx/fem/assemble_matrix_impl.h

```diff
@@ -41,7 +41,10 @@ void assemble_cells(la::MatSet<T> auto mat_set, mdspan2_t x_dofmap,
                     std::span<const std::int8_t> bc1, FEkernel<T> auto kernel,
                     std::span<const T> coeffs, int cstride,
                     std::span<const T> constants,
-                    std::span<const std::uint32_t> cell_info)
+                    std::span<const std::uint32_t> cell_info0,
+                    std::span<const std::uint32_t> cell_info1,
+                    const std::function<std::int32_t(std::int32_t)>& cell_map0,
+                    const std::function<std::int32_t(std::int32_t)>& cell_map1)
```

Comment:
Wouldn't the interface be simpler if we pass in `std::span<const std::int32_t>` instead? We already store the span inside entity maps in `entity_maps`.
I get that this means that we increase the memory of normal integration kernels with 1xnum_cells (as we could pass in the same span in both test and trial space). However, that is a relatively low cost.

Reply:
I already raised this idea with @jpdean. There are three lists we would need: `cells` - to get the geometry, and `cells_0` and `cells_1` which are mapped. For normal kernels, these would all be the same span, so no memory overhead.

Reply:
I'm happy to change to this approach if it is preferred

Reply:
The main issue with passing spans is that we'd need an "identity map" span for single domain problems, and it seems silly to have to store this. I tried to get around it by passing and empty span in this case and creating lambda functions where needed, but this is messy and not necessarily a good idea.

One approach might be to pass `entity_maps` to `create_form_factory` only, use the maps there, and for each integral, store the mapped cells along with the active cells. This data could then be passed to the assembler as a span, and in the case of single domain problems, we could just pass the same span twice. It also means we have contiguous access in the assembler with no need to map there. However, it might use more memory in certain cases, and it will also make [this](https://github.com/FEniCS/dolfinx/blob/main/cpp/dolfinx/fem/Form.h#L88-L94) (already pretty hefty) argument more complex.